File: sdk/common/src/main/java/ink/reactor/sdk/config/ConfigSection.java
package ink.reactor.sdk.config;


import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.function.BiConsumer;

public interface ConfigSection {
    String getName();

    Object get(final String key);
    <T> T get(final String key, final Class<T> clazz);
    <T> T getOrDefault(final String key, final T defaultValue);
    <T> List<T> getList(final String key, final Class<T> clazz);

    Collection<String> getKeys();
    Collection<?> getValues();

    Map<String, ?> getData();

    String getString(final String key);

    double getDouble(final String key);
    long getLong(final String key);
    int getInt(final String key);
    float getFloat(final String key);
    char getChar(final String key);
    boolean getBoolean(final String key);

    ConfigSection getSection(final String key);

    default List<String> getStringList(final String key) {
        return getList(key, String.class);
    }

    void forEachEntry(final BiConsumer<String, Object> consumer);

    Object set(final String key, final Object object);

    // Example: set("key1", "value", "key2", 25, "key3", "value3");
    default void set(final Object... keyAndValues) {
        if (keyAndValues.length % 2 != 0) {
            throw new RuntimeException("Key and values format: key, value, key, value.");
        }
        String keyString = null;
        for (Object object : keyAndValues) {
            if (keyString == null) {
                keyString = object.toString();
                continue;
            }
            set(keyString, object);
            keyString = null;
        }
    }

    boolean isEmpty();
}

File: sdk/common/src/main/kotlin/ink/reactor/sdk/config/ConfigSection.kt
package ink.reactor.sdk.config

interface ConfigSection {
    val name: String
    val data: MutableMap<String, Any?>

    operator fun get(key: String): Any? = data[key]
    operator fun set(key: String, value: Any?) { data[key] = value }

    infix fun String.to(value: Any?) {
        data[this] = value
    }

    fun getString(key: String): String? = get(key) as? String
    fun getInt(key: String): Int = (get(key) as? Number)?.toInt() ?: 0
    fun getLong(key: String): Long = (get(key) as? Number)?.toLong() ?: 0L
    fun getDouble(key: String): Double = (get(key) as? Number)?.toDouble() ?: 0.0
    fun getBoolean(key: String): Boolean = get(key) as? Boolean ?: false

    @Suppress("UNCHECKED_CAST")
    fun <T> getOrDefault(key: String, default: T): T {
        val value = data[key] ?: return default
        return try {
            value as T
        } catch (_: ClassCastException) {
            default
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun getSection(key: String): ConfigSection? {
        val sectionData = data[key] as? MutableMap<String, Any?> ?: return null
        return MapConfigSection(sectionData, key)
    }

    fun getOrCreateSection(key: String): ConfigSection {
        val existing = getSection(key);
        if (existing != null) {
            return existing
        }
        val newMap = LinkedHashMap<String, Any?>()
        data[key] = newMap
        return MapConfigSection(newMap, key)
    }

    fun <T : Any> delegate(default: T, key: String? = null) = ConfigDelegate(this, default, key)
}
